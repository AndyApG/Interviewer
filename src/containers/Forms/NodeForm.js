import React, { Component } from 'react';
import { reduxForm, Field } from 'redux-form';
import { required } from '../../utils/Validations';

import { Input, Form } from 'semantic-ui-react';

class ParticipantForm extends Component {
  textField = ({ input, label, meta: { touched, error }}) => {
    return (
      <Input label={label} {...input} />
    );
  }

  render() {
    const {
      props: {
        handleSubmit,
        submitButton
      },
      textField
    } = this;

    return (
      <Form onSubmit={handleSubmit}>
        <Form.Field>
          <Field name='fName' component={textField} label='First name' validate={[required]}/>
        </Form.Field>
        <Form.Field>
          <Field name='lName' component={textField} label='Last name' validate={[required]}/>
        </Form.Field>
        {submitButton}
      </Form>
    )
  }
}

ParticipantForm = reduxForm({
  form: 'node'
})(ParticipantForm);

export default ParticipantForm;
