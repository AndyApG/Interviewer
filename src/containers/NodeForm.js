import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { reset } from 'redux-form';
import PropTypes from 'prop-types';
import { pick, map } from 'lodash';
import { createSelector } from 'reselect';
import cx from 'classnames';

import { Button, ToggleInput } from '../ui/components';
import { actionCreators as modalActions } from '../ducks/modules/modals';
import { nodeAttributesProperty } from '../ducks/modules/network';
import { Form, FormWizard } from '../containers/';
import { Modal } from '../components/';
import { makeRehydrateFields } from '../selectors/forms';

const propNodeAttributes = (_, props) => props.node && props.node[nodeAttributesProperty];

const makePropFieldVariables = () =>
  createSelector(
    makeRehydrateFields(),
    fields => map(fields, 'name'),
  );

const makeGetPropFieldValues = () =>
  createSelector(
    makeRehydrateFields(),
    fields => Object.assign({}, ...fields.filter(field => field.value)
      .map(field => ({ [field.name]: field.value }))),
  );

const makeGetInitialValuesFromProps = () =>
  createSelector(
    makePropFieldVariables(),
    makeGetPropFieldValues(),
    propNodeAttributes,
    (fields, values, nodeAttributes) => ({ ...values, ...pick(nodeAttributes, fields) }),
  );

/**
  * Modal Node Form, than can handle new/editing of nodes
  * @extends Component
  */
class NodeForm extends Component {
  constructor(props) {
    super(props);
    this.modalRef = React.createRef();
    this.state = {
      addAnotherNode: false,
    };
  }

  onSubmit = (formData, dispatch, form) => {
    this.props.onSubmit(formData, dispatch, form);

    if (this.state.addAnotherNode) {
      this.props.resetValues(form.form);
      this.modalRef.current.scrollContentsToTop();
    } else {
      this.props.closeModal(this.props.name);
    }
  };

  onToggleClick = () => {
    this.setState({
      addAnotherNode: !this.state.addAnotherNode,
    });
  }

  render() {
    const {
      name,
      title,
      showAddAnotherToggle,
      useFullScreenForms,
    } = this.props;

    const modalClassNames = cx({ 'modal--fullscreen': useFullScreenForms });

    const formProps = {
      ...this.props,
      autoFocus: true,
      controls: [
        (showAddAnotherToggle && <ToggleInput
          key="toggleInput"
          name="addAnother"
          label="Add another?"
          checked={this.state.addAnotherNode}
          onCheck={this.onToggleClick}
          inline
        />),
        <Button key="submit" aria-label="Submit">Finished</Button>,
      ],
      form: name.toString(),
      onSubmit: this.onSubmit,
    };

    const formElement = useFullScreenForms ?
      (<FormWizard
        {...formProps}
      />) :
      (<Form
        {...formProps}
      />);

    return (
      <Modal name={name} title={title} className={modalClassNames} ref={this.modalRef}>
        {formElement}
      </Modal>
    );
  }
}

NodeForm.propTypes = {
  closeModal: PropTypes.func.isRequired,
  fields: PropTypes.array.isRequired,
  onSubmit: PropTypes.func.isRequired,
  initialValues: PropTypes.any.isRequired,
  name: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.symbol,
  ]).isRequired,
  title: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  entity: PropTypes.string.isRequired,
  node: PropTypes.object,
  openModal: PropTypes.func.isRequired,
  resetValues: PropTypes.func.isRequired,
  showAddAnotherToggle: PropTypes.bool,
  useFullScreenForms: PropTypes.bool.isRequired,
};

NodeForm.defaultProps = {
  showAddAnotherToggle: false,
  node: {},
};

function makeMapStateToProps() {
  const getInitialValuesFromProps = makeGetInitialValuesFromProps();

  return function mapStateToProps(state, props) {
    return {
      initialValues: getInitialValuesFromProps(state, props),
      useFullScreenForms: state.deviceSettings.useFullScreenForms,
    };
  };
}

function mapDispatchToProps(dispatch) {
  return {
    closeModal: bindActionCreators(modalActions.closeModal, dispatch),
    openModal: bindActionCreators(modalActions.openModal, dispatch),
    resetValues: bindActionCreators(reset, dispatch),
  };
}

export { NodeForm };

export default connect(makeMapStateToProps, mapDispatchToProps)(NodeForm);
