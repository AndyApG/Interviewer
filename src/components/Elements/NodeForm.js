/* eslint-disable no-shadow, react/no-unused-prop-types */

import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { map, pick } from 'lodash';
import { createSelector } from 'reselect';
import { Modal } from '../../containers/Elements';
import { Form } from '../../components/Elements';

const fieldNames = (_, props) => map(props.form.fields, 'name');
const node = (_, props) => props.node;

const initialValues = createSelector(
  fieldNames,
  node,
  (fieldNames, node) => pick(node, fieldNames),
);

/**
  * Modal Node Form
  */
const NodeForm = props => (
  <Modal name={props.modalName} title={props.form.title}>
    <Form
      {...props.form}
      initialValues={props.initialValues}
      form={props.form.formName}
      onSubmit={props.handleSubmit}
    />
  </Modal>
);

NodeForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  initialValues: PropTypes.any.isRequired,
  modalName: PropTypes.string.isRequired,
  form: PropTypes.any.isRequired,
  node: PropTypes.any,
};

NodeForm.defaultProps = {
  node: {},
};

function mapStateToProps(state, ownProps) {
  return {
    initialValues: initialValues(null, ownProps),
  };
}

export default connect(mapStateToProps)(NodeForm);
