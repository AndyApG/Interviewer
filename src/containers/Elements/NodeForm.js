import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { reset } from 'redux-form';
import PropTypes from 'prop-types';
import { pick, map } from 'lodash';
import { createSelector } from 'reselect';
import cx from 'classnames';

import { actionCreators as modalActions } from '../../ducks/modules/modals';
import { Form, FormWizard } from '../../containers/Elements';
import { Modal } from '../../components/Elements';
import { makeRehydrateFields } from '../../selectors/rehydrate';

const propNode = (_, props) => props.node;

const makePropFieldVariables = () =>
  createSelector(
    makeRehydrateFields(),
    fields => map(fields, 'name'),
  );

const makeGetInitialValuesFromProps = () =>
  createSelector(
    makePropFieldVariables(),
    propNode,
    (fields, node) => pick(node, fields),
  );

/**
  * Modal Node Form, than can handle new/editing of nodes
  * @extends Component
  */
class NodeForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      typeOfSubmit: 'normal',
    };
  }

  onSubmit = (formData, dispatch, form) => {
    this.close();
    this.props.handleSubmit(formData, dispatch, form);
    if (this.state.typeOfSubmit === 'continuous') {
      this.props.resetValues(form.form);
      this.props.openModal(this.props.name);
    }
  };

  continuousSubmit = () => {
    this.setState({
      typeOfSubmit: 'continuous',
    }, this.submit);
  };

  normalSubmit = () => {
    this.setState({
      typeOfSubmit: 'normal',
    }, this.submit);
  };


  close = () => {
    this.props.closeModal(this.props.name);
    this.setState({
      fieldIndex: 0,
    });
  };

  isLarge = () => window.matchMedia('screen and (min-device-aspect-ratio: 16/9)').matches;

  render() {
    const {
      title,
      fields,
      autoPopulate,
      name,
      addAnother,
      initialValues,
    } = this.props;

    const modalClassNames = cx({ 'modal--mobile': !this.isLarge() });

    const props = {};
    props.fields = fields;
    props.autoPopulate = autoPopulate;
    props.initialValues = initialValues;
    props.autoFocus = true;
    props.form = name.toString();
    props.onSubmit = this.onSubmit;
    props.addAnother = addAnother;
    props.continuousSubmit = this.continuousSubmit;
    props.normalSubmit = this.normalSubmit;

    const formElement = this.isLarge() ?
      (<Form
        {...props}
      />) :
      (<FormWizard
        {...props}
      />);

    return (
      <Modal name={name} title={title} close={this.close} className={modalClassNames}>
        {formElement}
      </Modal>
    );
  }
}

NodeForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  openModal: PropTypes.func.isRequired,
  resetValues: PropTypes.func.isRequired,
  initialValues: PropTypes.any.isRequired,
  name: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.symbol,
  ]).isRequired,
  title: PropTypes.string.isRequired,
  fields: PropTypes.array.isRequired,
  autoPopulate: PropTypes.func,
  node: PropTypes.any, // eslint-disable-line react/no-unused-prop-types
  addAnother: PropTypes.bool,
};

NodeForm.defaultProps = {
  addAnother: false,
  autoPopulate: () => {},
  node: {},
};

function makeMapStateToProps() {
  const getInitialValuesFromProps = makeGetInitialValuesFromProps();

  return function mapStateToProps(state, props) {
    return {
      initialValues: getInitialValuesFromProps(state, props),
    };
  };
}

function mapDispatchToProps(dispatch) {
  return {
    closeModal: bindActionCreators(modalActions.closeModal, dispatch),
    openModal: bindActionCreators(modalActions.openModal, dispatch),
    resetValues: bindActionCreators(reset, dispatch),
  };
}

export default connect(makeMapStateToProps, mapDispatchToProps)(NodeForm);
