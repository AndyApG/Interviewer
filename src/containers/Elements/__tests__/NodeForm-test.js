/* eslint-env jest */

import React from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { shallow, mount } from 'enzyme';
import NodeForm from '../NodeForm';

window.matchMedia = window.matchMedia || (() => ({
  matches: false, addListener: () => {}, removeListener: () => {},
}));

const node = {
  foo: 'bar',
};

const registry = {
  foo: {
    type: 'string',
    label: 'Foo',
  },
};

const mockStore = () =>
  createStore(
    () => (
      {
        protocol: {
          config: {
            registry,
          },
        },
        modals: [
          { name: 'baz', open: true },
        ],
      }
    ),
  );

const mockProps = {
  name: 'baz',
  title: 'The form title',
  handleSubmit: () => {},
  fields: [{
    variable: 'foo',
    component: 'TextInput',
  }],
};

describe('<NodeForm />', () => {
  it('should render', () => {
    const subject = shallow((
      <NodeForm {...mockProps} store={mockStore()} />
    ));

    expect(subject).toMatchSnapshot();
  });

  it('should render add another button', () => {
    const multipleControls = mount((
      <Provider store={mockStore()}>
        <NodeForm {...mockProps} addAnother />
      </Provider>
    ));
    const singularForm = mount((
      <Provider store={mockStore()}>
        <NodeForm {...mockProps} />
      </Provider>
    ));

    expect(multipleControls.find('form').find('button').length).toBe(2);
    expect(singularForm.find('form').find('button').length).toBe(1);
  });

  it('should render with prepopulated fields if provided', () => {
    const subject = mount((
      <Provider store={mockStore()}>
        <NodeForm {...mockProps} node={node} />
      </Provider>
    ));

    expect(subject.find('Form').first().prop('initialValues')).toEqual({ foo: 'bar' });
  });
});
