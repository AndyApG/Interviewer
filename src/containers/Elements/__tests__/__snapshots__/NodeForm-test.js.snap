// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NodeForm /> should render 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Connect(Modal)
    className="modal--fullscreen"
    name="baz"
    title="The form title"
  >
    <FormWizard
      addAnother={false}
      autoFocus={true}
      closeModal={[Function]}
      controls={
        Array [
          false,
          <Button
            aria-label="Submit"
            color=""
            content=""
            icon=""
            iconPosition="right"
            onClick={[Function]}
            size=""
          >
            Submit
          </Button>,
        ]
      }
      entity="node"
      fields={
        Array [
          Object {
            "component": "TextInput",
            "variable": "foo",
          },
        ]
      }
      form="baz"
      initialValues={Object {}}
      name="baz"
      node={Object {}}
      onSubmit={[Function]}
      openModal={[Function]}
      resetValues={[Function]}
      title="The form title"
      type="person"
    />
  </Connect(Modal)>,
  "nodes": Array [
    <Connect(Modal)
      className="modal--fullscreen"
      name="baz"
      title="The form title"
    >
      <FormWizard
        addAnother={false}
        autoFocus={true}
        closeModal={[Function]}
        controls={
          Array [
            false,
            <Button
              aria-label="Submit"
              color=""
              content=""
              icon=""
              iconPosition="right"
              onClick={[Function]}
              size=""
            >
              Submit
            </Button>,
          ]
        }
        entity="node"
        fields={
          Array [
            Object {
              "component": "TextInput",
              "variable": "foo",
            },
          ]
        }
        form="baz"
        initialValues={Object {}}
        name="baz"
        node={Object {}}
        onSubmit={[Function]}
        openModal={[Function]}
        resetValues={[Function]}
        title="The form title"
        type="person"
      />
    </Connect(Modal)>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <NodeForm
        addAnother={false}
        closeModal={[Function]}
        entity="node"
        fields={
          Array [
            Object {
              "component": "TextInput",
              "variable": "foo",
            },
          ]
        }
        initialValues={Object {}}
        name="baz"
        node={Object {}}
        onSubmit={[Function]}
        openModal={[Function]}
        resetValues={[Function]}
        title="The form title"
        type="person"
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": NodeForm {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "continuousSubmit": [Function],
        "isLarge": [Function],
        "normalSubmit": [Function],
        "onSubmit": [Function],
        "props": Object {
          "addAnother": false,
          "closeModal": [Function],
          "entity": "node",
          "fields": Array [
            Object {
              "component": "TextInput",
              "variable": "foo",
            },
          ],
          "initialValues": Object {},
          "name": "baz",
          "node": Object {},
          "onSubmit": [Function],
          "openModal": [Function],
          "resetValues": [Function],
          "title": "The form title",
          "type": "person",
        },
        "refs": Object {},
        "state": Object {
          "typeOfSubmit": "normal",
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Connect(Modal)
          className="modal--fullscreen"
          name="baz"
          title="The form title"
        >
          <FormWizard
            addAnother={false}
            autoFocus={true}
            closeModal={[Function]}
            controls={
              Array [
                false,
                <Button
                  aria-label="Submit"
                  color=""
                  content=""
                  icon=""
                  iconPosition="right"
                  onClick={[Function]}
                  size=""
                >
                  Submit
                </Button>,
              ]
            }
            entity="node"
            fields={
              Array [
                Object {
                  "component": "TextInput",
                  "variable": "foo",
                },
              ]
            }
            form="baz"
            initialValues={Object {}}
            name="baz"
            node={Object {}}
            onSubmit={[Function]}
            openModal={[Function]}
            resetValues={[Function]}
            title="The form title"
            type="person"
          />
        </Connect(Modal)>,
        "_debugID": 2,
        "_renderedOutput": <Connect(Modal)
          className="modal--fullscreen"
          name="baz"
          title="The form title"
        >
          <FormWizard
            addAnother={false}
            autoFocus={true}
            closeModal={[Function]}
            controls={
              Array [
                false,
                <Button
                  aria-label="Submit"
                  color=""
                  content=""
                  icon=""
                  iconPosition="right"
                  onClick={[Function]}
                  size=""
                >
                  Submit
                </Button>,
              ]
            }
            entity="node"
            fields={
              Array [
                Object {
                  "component": "TextInput",
                  "variable": "foo",
                },
              ]
            }
            form="baz"
            initialValues={Object {}}
            name="baz"
            node={Object {}}
            onSubmit={[Function]}
            openModal={[Function]}
            resetValues={[Function]}
            title="The form title"
            type="person"
          />
        </Connect(Modal)>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <NodeForm
    addAnother={false}
    closeModal={[Function]}
    entity="node"
    fields={
      Array [
        Object {
          "component": "TextInput",
          "variable": "foo",
        },
      ]
    }
    initialValues={Object {}}
    name="baz"
    node={Object {}}
    onSubmit={[Function]}
    openModal={[Function]}
    resetValues={[Function]}
    title="The form title"
    type="person"
  />,
}
`;
